### PR TITLE
Add new feature releated to Telegrambot

### PR DESCRIPTION
**Pull request by albertaleksieiev**
- Upload to google drive `Music Collection` folder, which you can configure in `UserProperty`
- Use Cookies from `UserProperty`
- Apply new `LocalAuth`, it fully compatible with current Spring Security

### CODE REVIEW

**Comment by albertaleksieiev** [comment]
@Denstden take a look at this, I cannot add you as a reviewers 😅

File: src/main/java/songbox/house/client/impl/VkClientImpl.java [hunk 1]
```diff
@@ -31,41 +35,37 @@
...
-        final Map<String, String> cookies = new HashMap<>();
+    @Override
+    public Map<String, String> getCookies() {
+        Map<String, String> cookies = new HashMap<>();
+        String cookieString = "";
+        try {
+            UserProperty userProperty = userPropertyService.getUserProperty();
+            cookieString = userProperty.getVkCookie();
+        } catch (Exception e) {
+            log.error("", e);
```
Comment (Denstden) [inline src/main/java/songbox/house/client/impl/VkClientImpl.java]: log.error("Can't load vk cookies", e)

File: src/main/java/songbox/house/client/impl/VkClientImpl.java [hunk 2]
```diff
@@ -31,41 +35,37 @@
...
+        }
+
+        if (TextUtils.isEmpty(cookieString)) {
```
Comment (Denstden) [inline src/main/java/songbox/house/client/impl/VkClientImpl.java]: I'd prefer to use StringUtils#isBlank

File: src/main/java/songbox/house/domain/entity/user/UserInfo.java [hunk 3]
```diff
@@ -66,4 +69,20 @@
             orphanRemoval = true
     )
     private List<YoutubePlaylist> youtubePlaylists = new ArrayList<>();
+
+    @OneToOne(cascade = ALL)
+    private UserProperty userProperty;
```
Comment (Denstden) [inline src/main/java/songbox/house/domain/entity/user/UserInfo.java:64]: Would be great to add lazy loading for this field

File: src/main/java/songbox/house/domain/entity/user/UserProperty.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    private Long userPropId;
+
+    @Column
+    private Boolean telegramBotUseGoogleDrive;
+
+    @Column(length = 2048)
+    private String vkCookie;
+
+    @OneToOne(cascade = ALL)
+    private MusicCollection defaultCollection;
```
Comment (Denstden) [inline src/main/java/songbox/house/domain/entity/user/UserProperty.java:28]: lazy

File: src/main/java/songbox/house/service/FrontendFriendlyService.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package songbox.house.service;
 
 import songbox.house.domain.dto.response.discogs.DiscogsReleaseDto;
+import songbox.house.domain.entity.MusicCollection;
```
Comment (Denstden) [inline src/main/java/songbox/house/service/FrontendFriendlyService.java]: unused

File: src/main/java/songbox/house/service/GoogleAuthenticationService.java [hunk 6]
```diff
@@ -8,7 +8,7 @@
 
     Drive getDrive();
 
-    String getRequestAccessUrl(String requestUrl);
+    String getRequestAccessUrl();
```
Comment (Denstden) [inline src/main/java/songbox/house/service/GoogleAuthenticationService.java]: REST project uses this method, I don't think these changes are really need for core. If you need for TG bot pass here some constant - do it in the telegram bot project.

OR, at least, fix REST project, taking into account these changes

File: src/main/java/songbox/house/service/UserService.java [hunk 7]
```diff
@@ -7,7 +7,8 @@
 
 public interface UserService {
 
-    UserInfo findByUserName(final String userName);
+    UserInfo findByUserName(String userName);
+    UserInfo findByUserNameOrCreate(String userName);
```
Comment (Denstden) [inline src/main/java/songbox/house/service/UserService.java]: What's the purpose of this method? We can't create a user without password

File: src/main/java/songbox/house/service/impl/GoogleAuthenticationServiceImpl.java [hunk 8]
```diff
@@ -51,6 +49,16 @@
     private static final GoogleClientSecrets GOOGLE_CLIENT_SECRETS = loadClientSecrets();
     private static final GoogleAuthorizationCodeFlow GOOGLE_AUTHORIZATION_CODE_FLOW = initFlow();
 
+    @Autowired
+    public GoogleAuthenticationServiceImpl(
+            @Value("${songbox.house.google.auth.redirectdomain}") String redirectDomain,
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/GoogleAuthenticationServiceImpl.java]: For REST project this prop is dynamic(it depends on request), but you made it static

File: src/main/java/songbox/house/service/impl/GoogleAuthenticationServiceImpl.java [hunk 9]
```diff
@@ -159,6 +168,11 @@ private GoogleClientSecrets getClientSecrets() {
     private void saveToken(GoogleTokenResponse googleTokenResponse, Long userId) {
         GoogleApiToken googleApiToken = createToken(googleTokenResponse, userId);
 
+        GoogleApiToken oldToken = googleApiTokenRepository.findByUserId(userId);
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/GoogleAuthenticationServiceImpl.java]: ofNullable(googleApiTokenRepository.findByUserId(userId)).ifPresent(googleApiTokenRepository::delete)

File: src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java [hunk 10]
```diff
@@ -23,28 +28,99 @@
...
 
     GoogleAuthenticationService authenticationService;
+    UserPropertyService userPropertyService;
 
-    @Override
-    public void upload(Track track) {
-        final Drive drive = authenticationService.getDrive();
+    private File getFolderOrCreate(Drive drive, String folderName, File parentFolder) throws IOException {
+        String escapedFolderName = folderName.replaceAll("'", "").replaceAll("\"", "");
+        String q;
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java:84]: To understand what does it mean I should read all related code

File: src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java [hunk 11]
```diff
@@ -23,28 +28,99 @@
...
+                    escapedFolderName,
+                    GOOGLE_DRIVE_MIMETYPE_FOLDER,
+                    parentFolder.getId());
+        }
+        FileList fileList = drive.files().list()
+                .setQ(q)
+                .execute();
+        List<File> files = fileList.getFiles();
+        if (files.isEmpty()) {
+            File fileMetadata = new File();
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java]: Move this block to the new method

File: src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java [hunk 12]
```diff
@@ -23,28 +28,99 @@
...
     private static final String MIME_TYPE = "audio/mpeg";
+    public static final String GOOGLE_DRIVE_MIMETYPE_FOLDER = "application/vnd.google-apps.folder";
 
     GoogleAuthenticationService authenticationService;
+    UserPropertyService userPropertyService;
 
-    @Override
-    public void upload(Track track) {
-        final Drive drive = authenticationService.getDrive();
+    private File getFolderOrCreate(Drive drive, String folderName, File parentFolder) throws IOException {
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java:82]: private methods should be after public

Comment (Denstden) [inline src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java:82]: the new code in this class is not clear enough for me, I hope you tested it and it works as it should

File: src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+@Service
+@AllArgsConstructor
+public class UserPropertyServiceImpl implements UserPropertyService {
+    private final UserInfoRepository userInfoRepository;
+    private final UserPropertyRepository userPropertyRepository;
+    private final UserService userService;
+
+    @Override
+    public UserProperty getUserProperty() {
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java]: getCurrentUserProperty?

File: src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        UserInfo currentUser = userService.getCurrentUser();
+        UserProperty userProperty = currentUser.getUserProperty();
+
+        if (currentUser.getUserProperty() == null) {
+            userProperty = new UserProperty();
+            currentUser.setUserProperty(userProperty);
+            userInfoRepository.save(currentUser);
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java:28]: What's the reason of saving empty object?

File: src/main/java/songbox/house/service/impl/UserServiceImpl.java [hunk 15]
```diff
@@ -109,6 +109,11 @@ public UserInfo findByUserName(final String userName) {
                 .orElseThrow(() -> new NotExistsException(format("User with name {0} not exists.", userName)));
     }
 
+    @Override
+    public UserInfo findByUserNameOrCreate(String userName) {
+        return null;
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/UserServiceImpl.java]: Shout it be here?

File: src/main/java/songbox/house/service/search/impl/FrontendFriendlyServiceImpl.java [hunk 16]
```diff
@@ -145,7 +142,11 @@ public void deleteFromCollection(Long id) {
         discogsReleaseRepository.deleteById(id);
     }
 
-    private MusicCollection getOrCreateMusicCollection() {
+    private MusicCollection getDefaultMusicCollection() {
+        MusicCollection defaultCollection = userPropertyService.getUserProperty().getDefaultCollection();
+        if (defaultCollection != null) {
+            return defaultCollection;
+        }
         return musicCollectionService.getOrCreate("frontend-friendly-collection");
```
Comment (Denstden) [inline src/main/java/songbox/house/service/search/impl/FrontendFriendlyServiceImpl.java:150]: return ofNullable(userPropertyService.getUserProperty().getDefaultCollection())
.orElseGet(() -> musicCollectionService.getOrCreate("frontend-friendly-collection"))

File: src/main/java/songbox/house/task/LoadVkAudioTask.java [hunk 17]
```diff
@@ -6,6 +6,7 @@
 import songbox.house.client.VkClient;
 import songbox.house.domain.entity.VkAudio;
 import songbox.house.util.JsonUtils;
+import songbox.house.util.ThreadChange;
```
Comment (Denstden) [inline src/main/java/songbox/house/task/LoadVkAudioTask.java:9]: is it used?

File: src/main/java/songbox/house/task/VkDownloadTrackTask.java [hunk 18]
```diff
@@ -5,6 +5,7 @@
 import com.mpatric.mp3agic.Mp3File;
 import lombok.extern.slf4j.Slf4j;
 import songbox.house.domain.entity.VkAudio;
+import songbox.house.util.ThreadChange;
```
Comment (Denstden) [inline src/main/java/songbox/house/task/VkDownloadTrackTask.java]: is it used?

File: src/main/resources/application.properties [hunk 19]
```diff
@@ -0,0 +1,54 @@
+# GENERAL
```
Comment (Denstden) [inline src/main/resources/application.properties]: IS this file really needed for core project?

File: src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java (hunk 13 above)
Comment (Denstden) [reply src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java]: Or pass user as parameter

File: src/test/java/songbox/house/threadchange/ThreadChangeTest.java [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    @Before
+    public void init() {
+        threadLocal.set(null);
+        executorService = ExecutorUtil.createExecutorService(4);
+    }
+
+    @Test
+    public void testThreadIsChangedAsRunnable() {
```
Comment (Denstden) [inline src/test/java/songbox/house/threadchange/ThreadChangeTest.java]: I think the prefix `test` in the test methods is redundant. My version for this test name -> `shouldPassThreadLocalParamsToNewRunnable`

File: src/test/java/songbox/house/threadchange/ThreadChangeTest.java [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+
+        ThreadChange.removeThreadChangeListener(threadLocalThreadChangeListener);
+    }
+
+    @Test
+    public void testThreadIsChangedAsTask() {
```
Comment (Denstden) [inline src/test/java/songbox/house/threadchange/ThreadChangeTest.java]: shouldPassThreadLocalParamsToNewCallable

File: src/main/java/songbox/house/service/UserService.java (hunk 7 above)
Comment (albertaleksieiev) [reply src/main/java/songbox/house/service/UserService.java]: I'll remove it

File: src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java (hunk 10 above)
Comment (albertaleksieiev) [reply src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java:84]: I'll add a comment

File: src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java (hunk 14 above)
Comment (albertaleksieiev) [reply src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java:28]: If there is no user property

File: src/main/java/songbox/house/task/VkDownloadTrackTask.java (hunk 18 above)
Comment (albertaleksieiev) [reply src/main/java/songbox/house/task/VkDownloadTrackTask.java]: This is just an **UNUSED** import, and yes it not used

File: src/test/java/songbox/house/threadchange/ThreadChangeTest.java (hunk 21 above)
Comment (albertaleksieiev) [reply src/test/java/songbox/house/threadchange/ThreadChangeTest.java]: for me `test` reads better than methods with should, why it should pass test?

File: src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java (hunk 14 above)
Comment (Denstden) [reply src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java:28]: I asked the reason, not the condition

File: src/test/java/songbox/house/threadchange/ThreadChangeTest.java (hunk 21 above)
Comment (Denstden) [reply src/test/java/songbox/house/threadchange/ThreadChangeTest.java]: For me it sounds like a tautology - you already have a class name with Test and annotation @Test. `should` because every test SHOULD pass, if it did not pass,- tested code is unstable and can't be used

File: src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java (hunk 14 above)
Comment (albertaleksieiev) [reply src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java:28]: > If there is no associated user property with current user

I hope it helps you

File: src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java (hunk 10 above)
Comment (Denstden) [reply src/main/java/songbox/house/service/impl/GoogleDriveServiceImpl.java:84]: Code should be self-documented. Just give the variables proper names

File: src/main/java/songbox/house/task/VkDownloadTrackTask.java (hunk 18 above)
Comment (Denstden) [reply src/main/java/songbox/house/task/VkDownloadTrackTask.java]: I means unused imports clog the code and are useless

File: src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java (hunk 14 above)
Comment (Denstden) [reply src/main/java/songbox/house/service/impl/UserPropertyServiceImpl.java:28]: this saving to the database is pointless and takes memory

File: src/main/java/songbox/house/service/impl/GoogleAuthenticationServiceImpl.java [hunk 22]
```diff
@@ -159,6 +164,9 @@ private GoogleClientSecrets getClientSecrets() {
     private void saveToken(GoogleTokenResponse googleTokenResponse, Long userId) {
         GoogleApiToken googleApiToken = createToken(googleTokenResponse, userId);
 
+        // Delete
```
Comment (Denstden) [inline src/main/java/songbox/house/service/impl/GoogleAuthenticationServiceImpl.java:167]: you can just move next line to a new method and name it `deleteIfExists(Long userId)` - then comment could be removed.

File: src/main/java/songbox/house/domain/entity/user/UserProperty.java [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    private Long userPropId;
+
+    @Column
+    private Boolean telegramBotUseGoogleDrive;
+
+    @Column(length = 2048)
+    private String vkCookie;
+
+    @OneToOne(cascade = ALL, fetch = FetchType.LAZY)
+    private MusicCollection defaultCollection;
```
Comment (Denstden) [inline src/main/java/songbox/house/domain/entity/user/UserProperty.java:28]: cascade = ALL means that if you remove collection -> User property will be removed as well(need to check, maybe it works reversed - if you remove property -> music collection will be removed as well)

Anyway, I think we don't need such behavior

**Review by Denstden — APPROVED** [review]
